pmm/queries.go

```diff
@@ -18,22 +18,367 @@
+			return nil, err
+		}
+		// Get parent_uuid again.
+		parentUUID, err = a.getAgentInstance(agentID)
+		if err != nil {
+			return nil, err
+		}
+	} else if err != nil {
+		return nil, err
+	}
+
+	// Check if related instance exists or try to re-use the existing one.
+	instance, err := a.getInstance(q.Name(), a.ServiceName, parentUUID)
+	if err == errNoInstance {
+		// Create new instance on QAN.
+		instance, err = a.createInstance(q.Name(), *info, parentUUID)
+		if err != nil {
+			return nil, err
+		}
+	} else if err != nil {
+		return nil, err
+	}
+
+	// Write instance config for qan-agent with real DSN.
+	instance.DSN = info.DSN
+	bytes, _ := json.MarshalIndent(instance, "", "    ")
+	if err := ioutil.WriteFile(fmt.Sprintf("%s/instance/%s.json", AgentBaseDir, instance.UUID), bytes, 0600); err != nil {
+		return nil, err
+	}
+
+	// Choose port.
+	port := 0
+	// Don't install service if we have already another one.
+	// 1 agent handles multiple instances for QAN.
+	if consulSvc == nil {
+		// Install and start service via platform service manager.
+		// We have to run agent before adding it to QAN.
+		svcConfig := &service.Config{
+			Name:        fmt.Sprintf("pmm-%s-queries-%d", q.Name(), port),
+			DisplayName: "PMM Query Analytics agent",
+			Description: "PMM Query Analytics agent",
+			Executable:  fmt.Sprintf("%s/bin/percona-qan-agent", AgentBaseDir),
+			Arguments:   a.Args,
+		}
+		if err := installService(svcConfig); err != nil {
+			return nil, err
+		}
+	} else {
+		port = consulSvc.Port
+		// Ensure qan-agent is started if service exists, otherwise it won't be enabled for QAN.
+		if err := startService(fmt.Sprintf("pmm-%s-queries-%d", q.Name(), port)); err != nil {
+			return nil, err
+		}
+	}
+
+	// Start QAN by associating instance with agent.
+	qanConfig := q.Config()
+	qanConfig.UUID = instance.UUID
+	qanConfig.Interval = 60
+	if err := a.startQAN(agentID, qanConfig); err != nil {
+		return nil, err
+	}
+
+	tags := []string{
+		fmt.Sprintf("alias_%s", a.ServiceName),
+	}
+	// For existing service, we append a new alias_ tag.
+	if consulSvc != nil {
+		tags = append(consulSvc.Tags, tags...)
+	}
+
+	// Add service to Consul.
+	serviceID := fmt.Sprintf("%s-%d", serviceType, port)
+	srv := consul.AgentService{
+		ID:      serviceID,
+		Service: serviceType,
+		Tags:    tags,
+		Port:    port,
+	}
+	reg := consul.CatalogRegistration{
+		Node:    a.Config.ClientName,
+		Address: a.Config.ClientAddress,
+		Service: &srv,
+	}
+	if _, err := a.consulAPI.Catalog().Register(&reg, nil); err != nil {
+		return nil, err
+	}
+
+	// Add info to Consul KV.
+	d := &consul.KVPair{
+		Key:   fmt.Sprintf("%s/%s/%s/dsn", a.Config.ClientName, serviceID, a.ServiceName),
+		Value: []byte(utils.SanitizeDSN(info.DSN)),
+	}
+	a.consulAPI.KV().Put(d, nil)
+	d = &consul.KVPair{
+		Key:   fmt.Sprintf("%s/%s/%s/qan_%s_uuid", a.Config.ClientName, serviceID, a.ServiceName, q.Name()),
+		Value: []byte(instance.UUID),
+	}
+	a.consulAPI.KV().Put(d, nil)
```

Comment:
Error return value of `(*github.com/percona/pmm-client/vendor/github.com/hashicorp/consul/api.KV).Put` is not checked